health/manager.go

```diff
@@ -0,0 +1,303 @@
+/*
+Package health implements a simple tool for health data + metrics collection.
+
+We defines three main parts of the health monitorin tool: manager, collector and writer.
+- Writer description is avilable in its own package.
+- Collector is stored in collector.go file. It provides a simple interface that allows you to collect
+different type of health data per target. All methods should have targetID as a parameter. Collector
+will automatically send data to health manager.
+- Manger is a heart of the health collection tool. It initialize all comunication channels and has
+a control over collector and writer. Manager keeps all data collected by collector, calculates every
+target health once per cycle and sends calculated data to writer.
+*/
+package health
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	logging "github.com/zenoss/zenoss-go-sdk/health/log"
+	"github.com/zenoss/zenoss-go-sdk/health/target"
+	w "github.com/zenoss/zenoss-go-sdk/health/writer"
+	"github.com/zenoss/zenoss-go-sdk/utils"
+)
+
+// Manager keeps all information about targets and provides a functionality
+// to initialize collector, writer and communication channels
+type Manager interface {
+	// Start method should initialize collector with it's configuration and
+	// define a method to send data to a writer
+	Start(ctx context.Context)
+	// AddTargets provides a simple interface to register monitored targets
+	AddTargets(targets []*target.Target)
+}
+
+// NewManager creates a new Manager instance with internal target registry.
+// You should init configuration and writer before and pass it here
+func NewManager(ctx context.Context, config *Config, writer w.HealthWriter) Manager {
+	healthReg := newHealthRegistry()
+
+	return &healthManager{
+		registry: healthReg,
+		config:   config,
+		writer:   writer,
+	}
+}
+
+// healthManager implements Manager interface
+type healthManager struct {
+	registry healthRegistry
+	config   *Config
+	writer   w.HealthWriter
+}
+
+func (hm *healthManager) Start(ctx context.Context) {
+	logging.SetLogLevel(hm.config.LogLevel)
+	measurements := make(chan *targetMeasurement)
+	initCollector(hm.config.CollectionCycle, measurements)
+	go hm.listenMeasurements(ctx, measurements)
+
+	healthIn := make(chan *target.Health)
+	go hm.healthForwarder(ctx, healthIn)
+	go hm.writer.Start(healthIn)
+}
+
+func (hm *healthManager) AddTargets(targets []*target.Target) {
+	hm.registry.lock()
+	defer hm.registry.unlock()
+
+	for _, target := range targets {
+		hm.registry.setRawHealthForTarget(newRawHealth(target, target.MetricIDs))
```

Comment:
I don't expect normal health collection to have a too many metrics etc. i.e. enough to cause problems with the memory based registry implementation. However, since the go-sdk and zdm etc. will be potentially used by a wider set (non-Zenoss) of developers, we should think about if & what guard rails to put in place to minimize or even preclude system wide issues due to a plugin. Limiting (could be reasonably large, but not infinite) the number of metrics & related concepts may be worth considering.

Reply:
Covered in ZING-17812